### DISCUSSION TITLE
Newbie Friendliness

### DISCUSSION THREAD

**Discussion by phil-barrett** [terjeio/ioSender#164] [Ideas]
This is in no way a criticism of ioSender.  I think it is an outstanding tool, best of breed.

But, in working with a number of customers, there are a few stumbling points. One that comes up somewhat often is the first run "config file not found" message. It is common for people to interpret that as an error.  Perhaps you could separate out missing from damaged and make it clear that missing is part of setting up.

Also, I am working with a local makerspace and in training new users, there are some stumbling points. Having the jog panel modal and not always visible is one.  I know it is not an easy change but it would be great to have the jog buttons always be visible and available. We  teach people to maximize the GCode Sender window so there is plenty of space available. Also, In my own use, I open/close the jog panel a LOT in each session. A floating panel as mention by others would be ok.

I would love to see outline as a visible panel below feed rate. We make the newbies do an air cut before running a real job. Outline is so much better - faster and doesn't require changing Z. 

With outline, I would like to see a warning if the Z axis is positioned outside of the workpiece outline. Or, perhaps make them move in side the outline before running. Sometimes clamps can be pretty tall and moving to the first point could easily hit something. I know you have a warning but the path to the first point is not immediately obvious.

Other issues have been mentioned - the router bit cone size in 3D view is one.

**Comment by phil-barrett** [comment]
A couple more:

Make Program/3D View/Console sticky across programs runs.  i.e. remember the last selection.

Show axes in 3D view before loading a GCode file. (this may have been asked for already).

**Comment by terjeio** [comment]
I guess I have to make a "newbie" version of the sender, what would the minimum required screen resolution be for that? The minimum window size is currently1020 x 700 pixels.
Should it support more than 3 axes? Max 4? The jog panel can grow quite wide so the layout might have to change depending on the screen width available if more than 4?.

**Reply by phil-barrett** [reply]
I honestly don't think you need a different version.  Display sizes have really grown in the last few years. At about $100 USD and up, most smaller PC displays these days are capable of at least 1600x900. The next tier up are all 1920x1080 or better. You can find smaller ones (11-12"/28-30cm) that are 1366x768. ​There are older laptops with less but most of the current 14" ones I looked are 1920x1080. For a truly lower bound, the tiny chromebooks and similar all appear to be 1366 x 768. An HP Stream Pro 11 G5 (11.6" display, 1366 x 768) is $270 USD. These are US prices and don't reflect tax. Maybe target 1366x768 min size.

If selecting a feature pushes the size above the display size, say it won't work. and revert to the fly out panel approach.

Yeah, I know the jog panel size is an issue but the vast majority of machines have 3 axes. Some have 4. I agree you should support all the machines out there.  Though, being insanely great for 90% and just ok for the remaining 10% would still be really good.

**Reply by terjeio** [reply]
> I honestly don't think you need a different version. 

I want that, at least for a start, as it will be far easier for me to start off with a larger canvas. The sender consists of many sub projects, the [main one](https://github.com/terjeio/ioSender/tree/master/GCode%20Sender/GCode%20Sender) is quite small and is where the UI is defined. In fact I have two other senders on top of the library projects, the public [config app](https://github.com/terjeio/ioSender/tree/master/Grbl%20Config%20App/Grbl%20Config%20App) and a private made for my CO2 laser.

WIth the current jog panel minimum screen width would have to be 1366 for 3-4 axes, six axes at least 1600. I recently made the buttons larger after a user request, he wanted that for touch control on a tablet. So I guess the buttons could be made smaller again...

**Reply by terjeio** [reply]
I have [uploaded a 1600 x900 version](http://www.io-engineering.com/downloads) with the jog panel visible.  Any ideas on how to arrange the controls in the UI?

**Reply by phil-barrett** [reply]
I like it! You are sooo fast! With a 4 axis machine, there is still room to add 2 more axes to the jog panel. I'd prefer the jog panel at the top right (vs bottom right) but that is minor. It would be nice to have any extra space given to the "view port" in the middle. Don't know if it is easily resizable. Here's a quick mockup of what I mean - wider  display area, jog panel to the top.  

I am also thinking that it might be nice to move some of the right [edit was: left] flyouts (goto and outline, specifically) into the space beneath program limits.  MPos would fit , too but that is likely to be less used.

The macro tab would be nice to tuck in somewhere, lower right hand side?  

I should be clear on the usage scenario I see - the default start up would have the display set to 3D View and have that as large as possible.  The people I've been teaching are drawn to it and it gives them a strong sense of what is going on.  The guys that set up machine at the Makerspace used OpenBuild's control app which is clumsy and fairly fragile.  I'm working get them to move to ioSender. Having everything visible in one screen is a big advantage.

![image](https://user-images.githubusercontent.com/16570958/144307445-bd9e6d8c-9463-498a-8907-fb9ba9c8f14d.png)

**Reply by phil-barrett** [reply]
I now understand you comment about number of axes.  Yes, limiting to 4 visible axes for screen space considerations would probably be OK. 

Also, I have not played with a rotary axis so not sure how that factor in.

**Reply by terjeio** [reply]
>  Having everything visible in one screen is a big advantage.

Well - I do not like it, and it does not fit on my laptop, so I guess it will be two versions - at least initially. BTW I never use UI jogging - keyboard jogging is my cup of tea. And the other flyouts are a keypress (ok two at the same time) away when I need them...

> The people I've been teaching are drawn to it and it gives them a strong sense of what is going on. The guys that set up machine at the Makerspace used OpenBuild's control app which is clumsy and fairly fragile.

I guess they use legacy Grbl then - for those I understand thate the UI jog panel is important.

It is not too much work to split the flyout controls in three as I have done for the jog panel, a base control with the content and two different wrappers for presentation. What will take more time will be to get scaling of the window content to behave correctly.

I am kind of quick now because I want a break from coding. Hopefully for grblHAL I will enter bug-fix/maintenance mode for a while. For ioSender I want version 2.0.36 published and then after a break look into improving the 3D view. I am thinking of calling this version _ioSender XL_ and making an edge version for getting feedback - initially without any fancy window scaling. If I can get scaling to work smoothly I may later merge the XL and "normal" versions.

**Reply by phil-barrett** [reply]
Thank you doing this. An XL version sounds good.  I do appreciate all your efforts.

And, you should definitely get more shop time! 

For the makerspace machine (a Millright V XL), I am hoping to move them to grblHAL. The MillRight Grbl+OpenBuilds is a bit flakey - 6 crashes in about 2 hrs of newbies using the machine.  Used ioSender with it but there seems to be comm corruption.  Have a new/better USB cable on the way.

**Reply by terjeio** [reply]
[XL version uploaded](http://www.io-engineering.com/downloads) - still work to do but first I want 2.0.36 out.

With this layout it is perhaps possible to go down to 1366x768? To be tested...

**Reply by molove** [reply]
1366 x 768 is good for me, I'm using an old laptop in my shop. My 2 cents worth

**Reply by calabr** [reply]
> [XL version uploaded](http://www.io-engineering.com/downloads) - still work to do but first I want 2.0.36 out.
> 
> With this layout it is perhaps possible to go down to 1366x768? To be tested...

XL version cannot be used with 1280x1024 screen.
Outline pannel consume to-much space. It's not so often needed.
Better to pin MPos tab on the left side instead.

Is it possible to set resizable middle window ?

**Reply by phil-barrett** [reply]
> Is it possible to set resizable middle window ?

In the regular version, if you maximize the app, the program/3d/console window (to which I think you are referring) gets resized. So, it is possible. Probably requires some work at startup.

**Reply by terjeio** [reply]
@molove @calabr I have reorganized the right hand side controls and reduced the minimum size to 1250x700, _Outline_ and _Goto_ controls will show up when there is space for them.

> Better to pin MPos tab on the left side instead.

As long as there is space to have flyout buttons on the right hand side it will go there.

> Is it possible to set resizable middle window ?

Not by itself, and I do not see any reason for making it resizable.

---

I have uploaded a [new edge version](http://www.io-engineering.com/downloads) with the reorganized controls, IMO this should be the basic configuration. Note that the sender is designed to support up to 6 axes, this put constraints on what can be displayed and when. I guess I should post a screen shot of how that looks like so that it becomes clear what amount screen real estate that I have available to play with at the minimum size.
For more advanced layout options I could make the app Win10 only, but since I have a feeling that many users use older computers to control their CNCs that would be a bad decision?

**Reply by phil-barrett** [reply]
Some feedback from a class I taught last night using ioSender XL. Very positive usage. Having the jog panel visible really cut down on the confusion. This is after we used the OpenBuilds Control for a previous class (it predated my involvement) so it was a low bar to clear. I'm still tinkering with the jog feed rate settings but think it works pretty well now. Had a couple of weird hangs but fewer than with OpenBuilds - I think it might be the Grbl Board (1.1f custom 4 axis build). Still working on convincing them that grblHAL is the way to go. My BoB could drop right in and give them an autosquaring gantry capability.

One small item came up - it would be good if Outline feed rate just kept the previous setting. It could be set in Settings: App but remembering the previous setting just seems more straight forward. The current default 500mm/min is a bit slow. Outline is a great tool, by the way, no need for a time wasting air cut. Last night, it prevented 3 students from wasting their piece of stock. We are drilling it into them as a key part of the Setup-Test-Run requirements.

**Reply by terjeio** [reply]
Great to hear.

> Had a couple of weird hangs

Are you able to reproduce any of those? Was it during running a job with no UI interaction?
I have not done any extensive testing with legacy Grbl so there may be more "bugs" to fix or workarounds to make for that. 

> it would be good if Outline feed rate just kept the previous setting

This will be added in the next edge version.

**Reply by phil-barrett** [reply]
No, the hangs were random.  They seemed to be associated with jogging. But, since jogging was about 90% of the interaction, that is not a very helpful observation.

When using the OpenBuilds sender, we had at least twice as many hangs - we had to hard reboot and rehome the machine each time. Pretty frustrating when it happens during a tool change and you have to re-zero the machine.

There was one very strange thing that happened.  Some how the Z origin changed while running a job.  And it changed up by 2.54mm  We were cutting with a depth of 0.1" (2.54mm).  Coincidence?  who knows.  I will try to reproduce this problem.  We've run about 10 jobs now with ioSender and only seen the problem once.

**Reply by terjeio** [reply]
> we had to hard reboot

I assume this is the controller board - it has either crashed or entered hard or soft limits triggered mode. Legacy grbl becomes unresponsive in all these cases even if was the intention to respond to [real time requests](https://github.com/gnea/grbl/blob/bfb67f0c7963fe3ce4aaf8a97f9009ea5a8db36e/grbl/protocol.c#L226-L238) on hard or soft limits triggered. The latter has been [fixed in grblHAL](https://github.com/grblHAL/core/blob/989d28a29ba88d105b0f29abf6d8e2753b07e894/protocol.c#L445-L459). My guess is that soft limits is enabled and a jog command triggers it. ioSender does not perform a limits check on UI jogging and I do not think I will add it. Keyboard jog commands are checked but is unstable with legacy grbl.

Legacy IMO Grbl can easily be fixed but is truly dead by now, switching to grblHAL is the only practical solution?

The Z-origin issue seems to be related to inches mode - but will be hard to track down unless it can be reproduced reliably.

**Reply by phil-barrett** [reply]
Yes, they have soft limits enabled (and hard, for homing). I am certain that it's a controller problem.  The mfr, MillRight, made their own 328P board and build of 1.1f.  They enabled just about every feature in Grbl including a 4th axis.  I wonder if the misbehavior comes from that. I tried to get the build specifics from MillRight and got nothing.  Biding my time to pitch the grblHAL conversion.

Yes, I am suspicious of the inch vs mm thing. Will keep an eye out for a reproducible case.

**Reply by terjeio** [reply]
New [edge versions uploaded](http://www.io-engineering.com/downloads) - jog distance in G20 mode seems to work now, also the UI jog panel is no longer disabled while in JOG state allowing multiple clicks on the jog buttons and the option to cancel an ongoing jog. Outline feedrate is remembered accross sessions. Translations has been incorporated in the XL version.

**Reply by phil-barrett** [reply]
Excellent!

**Comment by willtel** [comment]
Hi Terje,
Are You ever going to Release (Public) your *config app* for private made
for your CO2 laser?
I am in the process of building a 1500 x 1500 mm X, (Y + B Ganged ) , Z__
Gantry Router with Changeable Heads (a)  Kress Spindle Head + (b)  CO2
Laser Attachment Head.
I am still in the Hardware building stage of the TABLE and ROUTER based
upon the Open Builds KYO_SPHINX and the Open Builds C-Beam Sphinx by Mark
Carew for cutting and Engraving PERSPEX, but I find my age (75YO) is
slowing me down, I am doing my own Electronics.

Thus I would be very interested in the ioSENDER with CO2 LASER SUPPORT.

**Reply by terjeio** [reply]
There is pretty good laser support already and there are [laser specific plugins](https://github.com/grblHAL/Plugins_laser) available for some grblHAL based controllers. The custom sender replaces the right hand side with different controls that allows binding profiles to tools for selecting PPI (pulses per inch) and pulse width - this needs the PPI plugin. Coolant control is also enhanced, this requires the coolant plugin:

![image](https://user-images.githubusercontent.com/20260062/144292099-467c609b-4d7d-4b2c-b7c1-afa157d90ca4.png)

I can release it, but I need to get some time on my own in the workshop first - I have not been there in a long time. Too busy with coding...

**Reply by willtel** [reply]
YES I like it, Your ioSENDER + grblHAL is the BEST SOLUTION I have come across, the ioSENDER_CO2 LASER Version is not urgent, I have still got a lot of work to do my end.
Thanking You for all your help.

**Comment by molove** [comment]
Another newbie request, is it possible to be able to fire up IOSender whilst not connected to the device running GrblHal?
I would like to be able to pore over the GUI to familiarise myself with it, my CNC lives in my shed, which I'd rather not sit in too much if possible.
It's not majorly important, as I can always flash Grbl or GrblHal to another device and plug that in when I want to fire up IOSender

**Reply by terjeio** [reply]
> Another newbie request, is it possible to be able to fire up IOSender whilst not connected to the device running GrblHal?

No, and I have no plans to change that. I may allow switching to another controller later on.
The main reason is that the sender configures itself from data requested from the controller, switching tabs may even rerequest data to be 100% sure about the current controller status (there could be a MPG changing that without the sender knowing).

> It's not majorly important, as I can always flash Grbl or GrblHal to another device and plug that in when I want to fire up IOSender

An option is to use the [Simulator](https://github.com/grblHAL/Simulator) with a network connection, this compiles and runs under Linux - I have not built it for Windows yet.

**Comment by phil-barrett** [comment]
Another item for newbies.  When the board USB is working (the PC sees a connection) but grblHAL has crashed, ioSender comes up but shows all possible tabs greyed out.  This confuses people.  Perhaps a message - "grblHAL connected but not responding, check configuration"  or something similar would help them look in the right places.  This came up because a customer enabled eeprom in my_macine.h without a chip in place. I knew what it was right away but he thought it was an ioSender problem.
